<a href="https://colab.research.google.com/github/YunilaPutmasari/Machine_Learning-2024/blob/main/TugasML_9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tugas
- Lakukan klasifikasi pada data MNIST dengan menggunakan model ANN

- Anda diperbolehkan melakukan eksplorasi terhadap,

  - Metode pra pengolahan
  - Pemilihan fitur
  -mArsitektur ANN
  - Fungsi Aktiviasi

- ANN diimplementasikan dengan menggunakan tensorflow.
- DIKERJAKAN SECARA BERKELOMPOK
- JELASKAN HASIL YANG ANDA DAPATKAN,
  - AKURASI
  - CONFUSION MATRIX
  - KONFIGURASI MODEL --> MULAI DARI PRA PENGOLAHAN SAMPAI ARSITEKTUR ANN

1. Persiapan Data dan Pra-Pengolahan

In [7]:
pip show tensorflow

Name: tensorflow
Version: 2.17.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: absl-py, astunparse, flatbuffers, gast, google-pasta, grpcio, h5py, keras, libclang, ml-dtypes, numpy, opt-einsum, packaging, protobuf, requests, setuptools, six, tensorboard, tensorflow-io-gcs-filesystem, termcolor, typing-extensions, wrapt
Required-by: dopamine_rl, tf_keras


In [2]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

# Load dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalisasi
x_train, x_test = x_train / 255.0, x_test / 255.0

# Reshaping
x_train = x_train.reshape(-1, 784)
x_test = x_test.reshape(-1, 784)

# One-hot encoding labels
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)


Berikut adalah ringkasan dari penjelasan yang diberikan:

### Import Library
Kode ini mengimpor TensorFlow dan modul yang diperlukan untuk memuat dataset MNIST serta melakukan one-hot encoding pada label.

### Memuat Dataset
`mnist.load_data()` digunakan untuk memuat dataset MNIST, yang terbagi menjadi data pelatihan (`x_train`, `y_train`) dan data pengujian (`x_test`, `y_test`).

### Normalisasi
Gambar dinormalisasi dari rentang 0-255 menjadi 0.0-1.0 dengan membagi nilai piksel dengan 255.0, yang membantu dalam pelatihan model.

### Reshaping
Gambar berukuran 28x28 piksel diubah menjadi vektor satu dimensi (784) untuk mempermudah pemrosesan oleh model.

### One-hot Encoding
Label angka (0-9) diubah menjadi format one-hot encoding menggunakan `to_categorical()`, misalnya label '3' menjadi `[0, 0, 0, 1, 0, 0, 0, 0, 0, 0]`, yang berguna untuk klasifikasi.

Secara keseluruhan, kode ini mempersiapkan dataset MNIST untuk pelatihan model dengan normalisasi, reshaping, dan konversi label.

2. Pemilihan Fitur

In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential([
    Dense(512, activation='relu', input_shape=(784,)),
    Dense(256, activation='relu'),
    Dense(10, activation='softmax')
])


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Berikut adalah ringkasan penjelasan tentang model jaringan saraf menggunakan Keras dalam TensorFlow:

1. **Import Model dan Layer**:
   - `Sequential`: Kelas untuk membuat model jaringan saraf secara berurutan.
   - `Dense`: Layer penuh (fully connected layer) yang menghubungkan semua neuron dari layer sebelumnya.

2. **Membuat Model**:
   - `model = Sequential([...])`: Instance model Sequential dengan layer-layer yang ditambahkan.

3. **Layer Pertama**:
   - `Dense(512, activation='relu', input_shape=(784,))`:
     - **512** neuron dengan fungsi aktivasi ReLU dan input bentuk 784 (vektor gambar).

4. **Layer Kedua**:
   - `Dense(256, activation='relu')`:
     - Layer tersembunyi kedua dengan 256 neuron dan fungsi aktivasi ReLU.

5. **Layer Output**:
   - `Dense(10, activation='softmax')`:
     - **10** neuron (untuk kelas 0-9) dengan fungsi aktivasi softmax untuk menghasilkan probabilitas kelas.

Model ini terdiri dari dua layer tersembunyi (ReLU) dan satu layer output (softmax) untuk klasifikasi data MNIST.

4. Kompilasi Model

In [4]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


Berikut adalah ringkasan penjelasan tentang `model.compile()` untuk model jaringan saraf:

1. **Optimizer**:
   - `optimizer='adam'`: Menggunakan algoritma optimisasi Adam, yang efisien dan cepat dalam konvergensi dengan mengadaptasi laju pembelajaran berdasarkan gradien.

2. **Loss Function**:
   - `loss='categorical_crossentropy'`: Menggunakan fungsi loss categorical crossentropy, yang umum untuk klasifikasi multi-kelas dan mengukur kesesuaian antara prediksi model dan label one-hot encoding.

3. **Metrics**:
   - `metrics=['accuracy']`: Menetapkan akurasi sebagai metrik evaluasi untuk memantau kinerja model selama pelatihan dan pengujian.

Secara keseluruhan, kode ini menyiapkan model dengan optimizer, fungsi loss, dan metrik evaluasi yang sesuai.

5. Pelatihan Model

In [5]:
history = model.fit(x_train, y_train, epochs=10, validation_split=0.2, batch_size=32)


Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 29s 18ms/step - accuracy: 0.8924 - loss: 0.3530 - val_accuracy: 0.9620 - val_loss: 0.1240
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 15s 10ms/step - accuracy: 0.9732 - loss: 0.0878 - val_accuracy: 0.9683 - val_loss: 0.1015
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 15s 10ms/step - accuracy: 0.9836 - loss: 0.0525 - val_accuracy: 0.9696 - val_loss: 0.1062
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 22s 11ms/step - accuracy: 0.9868 - loss: 0.0405 - val_accuracy: 0.9765 - val_loss: 0.0909
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 15s 10ms/step - accuracy: 0.9900 - loss: 0.0286 - val_accuracy: 0.9707 - val_loss: 0.1032
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 20s 10ms/step - accuracy: 0.9916 - loss: 0.0248 - val_accuracy: 0.9743 - val_loss: 0.1104
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 15s 10ms/step - accuracy: 0.9931 - loss: 0.0219 - val_accuracy: 0.9674 - val_loss: 0.1432
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 20s 9ms/step - accuracy: 0.9925 - 

Berikut adalah ringkasan penjelasan tentang `model.fit()` untuk melatih model jaringan saraf:

1. **Data Pelatihan**:
   - `x_train`: Data input (gambar MNIST yang dinormalisasi dan di-reshape).
   - `y_train`: Label target dalam format one-hot encoding.

2. **Epochs**:
   - `epochs=10`: Model dilatih selama 10 iterasi penuh terhadap dataset.

3. **Validation Split**:
   - `validation_split=0.2`: Menggunakan 20% dari data pelatihan untuk validasi, memantau kinerja model dan mencegah overfitting.

4. **Batch Size**:
   - `batch_size=32`: Menetapkan ukuran batch, memproses 32 contoh data sekaligus sebelum memperbarui bobot.

5. **Return Value**:
   - `history`: Menyimpan hasil pelatihan, termasuk loss dan akurasi untuk analisis lebih lanjut.

Secara keseluruhan, kode ini melatih model dengan dataset MNIST selama 10 epoch, menggunakan validasi dan ukuran batch 32.

6. Evaluasi Model

In [6]:
# Evaluasi akurasi pada data uji
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy}")

# Confusion matrix
from sklearn.metrics import confusion_matrix
import numpy as np

# Prediksi pada data uji
y_pred = model.predict(x_test)
y_pred_classes = np.argmax(y_pred, axis=1)
y_true = np.argmax(y_test, axis=1)

# Buat confusion matrix
cm = confusion_matrix(y_true, y_pred_classes)
print("Confusion Matrix:")
print(cm)


313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9702 - loss: 0.1461
Test Accuracy: 0.9764999747276306
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step
Confusion Matrix:
[[ 974    0    0    1    0    0    3    0    1    1]
 [   2 1124    0    1    0    0    2    0    5    1]
 [   4    6  996    9    1    0    4    5    7    0]
 [   0    0    1  988    0    7    0    3    1   10]
 [   3    0    1    1  943    0    9    2    2   21]
 [   2    0    0    6    0  873    4    0    1    6]
 [   3    2    0    1    1    8  942    0    1    0]
 [   1    3    3    1    1    0    0 1000    4   15]
 [   3    1    4    8    5    5    1    4  930   13]
 [   2    2    0    1    5    3    1    0    0  995]]


Kode ini melakukan evaluasi akurasi model pada data uji dan menghasilkan matriks kebingungan untuk melihat detail klasifikasi.

### Penjelasan Kode
1. **Evaluasi Akurasi pada Data Uji**:
   ```python
   test_loss, test_accuracy = model.evaluate(x_test, y_test)
   print(f"Test Accuracy: {test_accuracy}")
   ```
   - `model.evaluate(x_test, y_test)`: Menghitung *loss* dan akurasi model pada data uji (`x_test` dan `y_test`).
   - Hasil akurasi pada data uji dicetak untuk memberikan informasi mengenai seberapa baik model memprediksi data yang belum pernah dilihat.

2. **Confusion Matrix (Matriks Kebingungan)**:
   ```python
   from sklearn.metrics import confusion_matrix
   import numpy as np
   ```
   - Mengimpor modul `confusion_matrix` dari `sklearn.metrics` dan `numpy` untuk membantu memproses data.

3. **Prediksi pada Data Uji**:
   ```python
   y_pred = model.predict(x_test)
   y_pred_classes = np.argmax(y_pred, axis=1)
   y_true = np.argmax(y_test, axis=1)
   ```
   - `model.predict(x_test)`: Menghasilkan prediksi probabilitas untuk setiap kelas pada data uji.
   - `np.argmax(y_pred, axis=1)`: Mengambil kelas dengan probabilitas tertinggi untuk setiap sampel, menghasilkan kelas yang diprediksi (`y_pred_classes`).
   - `np.argmax(y_test, axis=1)`: Mengubah label one-hot pada data uji (`y_test`) menjadi bentuk kelas asli (`y_true`).

4. **Membangun Confusion Matrix**:
   ```python
   cm = confusion_matrix(y_true, y_pred_classes)
   print("Confusion Matrix:")
   print(cm)
   ```
   - `confusion_matrix(y_true, y_pred_classes)`: Menghasilkan matriks kebingungan untuk membandingkan prediksi model dengan label asli.
   - Matriks ini menunjukkan distribusi hasil prediksi model, membantu menganalisis jenis kesalahan yang sering terjadi.

# Penjelasan hasil

Berikut adalah ringkasan hasil evaluasi model, mencakup akurasi, matriks kebingungan, dan konfigurasi model dari pra-pengolahan hingga arsitektur ANN:

### 1. Akurasi
- **Test Accuracy: 0.9765 (97.65%)**: Model berhasil mengklasifikasikan 97.65% dari data uji dengan benar, menunjukkan performa yang sangat baik dalam mengenali angka tulisan tangan pada dataset MNIST.

### 2. Confusion Matrix
- **Diagonal Utama**: Menunjukkan jumlah prediksi yang benar untuk setiap kelas. Sebagian besar kelas memiliki angka tinggi, menandakan akurasi yang baik.
- **Kesalahan Klasifikasi**: Kelas yang paling sering salah adalah angka '8', yang sering terklasifikasi sebagai '3', '4', atau '5'. Model sangat baik dalam mengenali angka '7' dan '1'.
  
### 3. Konfigurasi Model
#### Pra-Pengolahan Data
- **Normalisasi**: Nilai piksel dinormalisasi ke rentang [0, 1].
- **Reshaping**: Gambar diubah dari 28x28 menjadi vektor 784.
- **One-Hot Encoding**: Label kelas diubah ke format one-hot.

#### Arsitektur Model ANN
1. **Layer Input**:
   - **Dense(512, activation='relu')**: Layer pertama dengan 512 neuron.
  
2. **Layer Tersembunyi**:
   - **Dense(256, activation='relu')**: Layer kedua dengan 256 neuron.

3. **Layer Output**:
   - **Dense(10, activation='softmax')**: Layer output dengan 10 neuron, memberikan probabilitas untuk setiap kelas.

#### Proses Pelatihan
- **Optimizer**: Adam untuk mempercepat konvergensi.
- **Loss Function**: Categorical crossentropy untuk klasifikasi multi-kelas.
- **Epochs**: 10 epoch dengan 20% data untuk validasi.
- **Batch Size**: 32 untuk efisiensi pemrosesan.

### Kesimpulan
Model menunjukkan akurasi di atas 97%, dengan matriks kebingungan mengungkapkan wawasan tentang kesalahan klasifikasi. Konfigurasi model yang baik dan pra-pengolahan data yang tepat berkontribusi pada performa yang mengesankan.